add-schema.sh : Update formatting of the lines being added to the pom.xml file

ianwallen opened the pull request:
Update formatting of the lines being added to the pom.xml file so that they fit better with the default geonetwork pom files.

josegar74 commented:
@ianwallen at least in OSX doesn't seem to make any difference. There's no indentation in the lines added.

ianwallen commented:
@josegar74  It is odd that it would not work on Mac - I thought it would be similar to linux. Is it possible that you were not comparing spaces?

I have tested it with GitBash on Windows system and I have also tested it on Linux Mint.

I tested it running the following command 

`./add-schema.sh iso19139.ca.HNAP https://github.com/metadata101/iso19139.ca.HNAP 3.10.x`

And then I checked the **web/pom.xml**

Before this pull request I get the following.

```
$ git diff web/pom.xml
diff --git a/web/pom.xml b/web/pom.xml
index 7d00ffaed8..6122603d89 100644
--- a/web/pom.xml
+++ b/web/pom.xml
@@ -515,6 +515,11 @@
       <artifactId>schema-iso19139</artifactId>
       <version>${gn.schemas.version}</version>
     </dependency>
+<dependency>
+<groupId>${project.groupId}</groupId>
+<artifactId>schema-iso19139.ca.HNAP</artifactId>
+<version>${gn.schemas.version}</version>
+</dependency>
     <dependency>
       <groupId>${project.groupId}</groupId>
       <artifactId>schema-iso19110</artifactId>
@@ -743,6 +748,10 @@
                   <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins
                   </targetPath>
                 </resource>
+ <resource>
+    <directory>${project.basedir}/../schemas/iso19139.ca.HNAP/src/main/plugin</directory>
+    <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins</targetPath>
+  </resource>
                 <resource>
                   <directory>${project.basedir}/../schemas/iso19110/src/main/plugin</directory>
                   <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins
$
```


After this change has been applied I get the following.

```
$ git diff web/pom.xml
diff --git a/web/pom.xml b/web/pom.xml
index 7d00ffaed8..8cc1d2f6fd 100644
--- a/web/pom.xml
+++ b/web/pom.xml
@@ -515,6 +515,11 @@
       <artifactId>schema-iso19139</artifactId>
       <version>${gn.schemas.version}</version>
     </dependency>
+    <dependency>
+      <groupId>${project.groupId}</groupId>
+      <artifactId>schema-iso19139.ca.HNAP</artifactId>
+      <version>${gn.schemas.version}</version>
+    </dependency>
     <dependency>
       <groupId>${project.groupId}</groupId>
       <artifactId>schema-iso19110</artifactId>
@@ -743,6 +748,10 @@
                   <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins
                   </targetPath>
                 </resource>
+                <resource>
+                  <directory>${project.basedir}/../schemas/iso19139.ca.HNAP/src/main/plugin</directory>
+                  <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins</targetPath>
+                </resource>
                 <resource>
                   <directory>${project.basedir}/../schemas/iso19110/src/main/plugin</directory>
                   <targetPath>${basedir}/src/main/webapp/WEB-INF/data/config/schema_plugins
$
```

As you can see from these results the indentation is cleaner with the new commits applied.

josegar74 commented:
@ianwallen I'll recheck, I opened the files in IntelliJ, but the text for the snippets was aligned to the left.

fxprunayre commented:
Works fine for me.

juanluisrp commented:
For me, using MacOS Catalina and ZSH shell, it doesn't set the indentation as expected:

![image](https://user-images.githubusercontent.com/826920/80456769-75700d00-892e-11ea-91f3-8b1fa937f001.png)

juanluisrp commented:
This works fine for me:
```bash
  sed $sedopt "${insertLine} a\\
\ \ \ \ <dependency>\\
\ \ \ \ \ \ <groupId>${projectGroupId}</groupId>\\
\ \ \ \ \ \ <artifactId>schema-${schema}</artifactId>\\
\ \ \ \ \ \ <version>${gnSchemasVersion}</version>\\
\ \ \ \ </dependency>
  " web/pom.xml
```

produces:
```xml
    <dependency>
      <groupId>${project.groupId}</groupId>
      <artifactId>schema-dcat-ap1.1</artifactId>
      <version>${gn.schemas.version}</version>
    </dependency>
```

juanluisrp commented:
Tested escaping the spaces in Ubuntu's Bash shell and works OK too.

ianwallen commented:
According to the following
https://scriptingosx.com/2019/06/moving-to-zsh/
Mac are using a really old version of bash.  Even if you run it from zsh - it should still run bash at is has "#!/bin/bash" at the top of the file.

I have modified the code so that it reads the sed script from the stdin instead of a multiline string.
Hopefully this will solve the problem without having to add all those escape characters.

Could someone test these new changes on the mac to see if it resolves the issue.  If not then I can add all the escape characters

Thank you

ianwallen commented:
@juanluisrp - would you be able to test the lastest changes to see if the suggested changes work. I would prefer using stdin instead of multiple escape characters. Unfortunately, I don't have a mac for me to test the new changes.

Thank you

juanluisrp commented:
No, that doesn't work
```
Adding schema dcat-ap1.1 to schemas/pom.xml
sed: -: No such file or directory
Adding schema dcat-ap1.1 dependency to web/pom.xml
sed: -: No such file or directory
Adding schema dcat-ap1.1 resources to web/pom.xml
sed: -: No such file or directory
```

juanluisrp commented:
The version of Bash I have is `3.2.57
```
/bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin19)
Copyright (C) 2007 Free Software Foundation, Inc.
```

ianwallen commented:
@juanluisrp 
Thank you for testing.

I was able to [install that version of bash on Linux](https://stackoverflow.com/questions/34805372/how-can-i-test-my-bash-script-on-older-versions-of-bashGNU%20bash,%20version%203.2.57)

```
$ ./bash -version
GNU bash, version 3.2.57(1)-release (x86_64-unknown-linux-gnu)
Copyright (C) 2007 Free Software Foundation, Inc.
```

With that version of bash it was working correctly on my system.
I noticed that your error is related to "sed". Maybe your version of sed does not support stdin for the script.
Could you run `sed --version` to get the version. Mine is version 4.4

```
sed --version
sed (GNU sed) 4.4
```

According to [the following](https://unix.stackexchange.com/questions/45591/using-a-here-doc-for-sed-and-a-file) there are other ways of passing the stdin to sed maybe you can try the following commands and let me know if any of them work.

```
#create temporary file to be used in the tests.
echo "this is a test" > /tmp/test.txt

# This will probably fail as it seem that it does not support "-" for stdin
sed -f - /tmp/test.txt << SED_SCRIPT
    1 a\\
    This\\
      Is\\
        A\\
          Test
SED_SCRIPT

#This may work but may still have indentation issues?
sed -f /dev/stdin /tmp/test.txt << SED_SCRIPT
    1 a\\
    This\\
      Is\\
        A\\
          Test
SED_SCRIPT

#This may work and have proper indentation?
sed -f /dev/stdin /tmp/test.txt <<-SED_SCRIPT
    1 a\\
    This\\
      Is\\
        A\\
          Test
SED_SCRIPT
```

I get the following for all these tests

```
this is a test
    This
      Is
        A
          Test
```

josegar74 commented:
@ianwallen, I think if you tested in Linux / Windows and works ok, it's fine to merge.  

`sed` in OSX has some differences with "standard" `sed`, see https://unix.stackexchange.com/questions/13711/differences-between-sed-on-mac-osx-and-other-standard-sed

ianwallen commented:
Wow! "sed" from 2005! That is old.
If we merge the initial commit that I had done, it would work for Windows/Linux but the indent would not be applied to OSX.  It we apply the current changes, it will not work at all on OSX.

So we should make a change to have OSX work again. Even if the indentation is not in place.
Since we should apply one last (hopefully) change - it would be nice if someone could run the commands above on OSX.  If I just need to changing "-f -"  to "-f /dev/stdin" then that should be the better fix. If they all fail, we can go back to the initial commit.

Sorry - I know you are busy and this was supposed to be a simple fix for the spacing.  If I would have know it was going to cause these issues, I would not have submitted the change.

ianwallen commented:
@juanluisrp, @josegar74 
I believe the last change that I have made should work on mac, linux and windows(via gitbash). 
Please test when you have time.

I incorporated Juan's change by escaping the space. However I found out that I did not need to escape them all - I only had to escape the first one.
https://www.grymoire.com/Unix/Sed.html#uh-43

I also fixed the issue with mac having to use /dev/stdin instead of "-"
This is what was causing the error.
      sed: -: No such file or directory
